### PR TITLE
feature/ft-814 service_person

### PR DESCRIPTION
**Pull request by KevinWeirNZ**
#1  [FT-814] Hoodoo Training - Step 4 - Build a Hoodoo service that interacts with a Postgres database.
Jira: [FT-814](https://loyaltynz.atlassian.net/browse/FT-814)

What?
* Following the defined [Loyalty instructions](http://loyaltynz.github.io/hoodoo/guides_0300_active_record.html) to build a Hoodoo service that interacts with a Postgres database.

Why?
* To gain experience of building a realistic full-stack API service.

### CODE REVIEW

File: db/migrate/20211130085736_add_index_to_name.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class AddIndexToName < ActiveRecord::Migration[5.1]
+  def up
+    # add_index :people, :name
+  end
+
+  def down
+    # drop_index :people, :name
+  end
+end
```
Comment (davidoram) [inline db/migrate/20211130085736_add_index_to_name.rb:9]: This file is not doing anything - best to delete it

File: service/implementations/person_implementation.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
+class PersonImplementation < Hoodoo::Services::Implementation
+
+  # def show( context )
+  #   person = Person.acquire_in!( context )
+  #   return if context.response.halt_processing?
+  #   context.response.set_resources( render_in( context, person ) )
+  # end
```
Comment (davidoram) [inline service/implementations/person_implementation.rb]: Can you delete this, rather than comment out.  We would look in git history to see previous versions of a file.

File: service/implementations/person_implementation.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    person = Person.acquire_in( context )
+    if person.nil?
+      context.response.not_found( context.request.ident )
+    else
+      context.response.set_resource( render_in( context, person ) )
+    end
+  end
+
+  def list( context )
+    # context.request.list.limitv= 1000 if context.request.list.limit > 1000 # Limit the size of page size a caller can request.
```
Comment (davidoram) [inline service/implementations/person_implementation.rb]: Can you guess why this might be useful to be kept in?
```suggestion
    context.request.list.limit= 1000 if context.request.list.limit > 1000 # Limit the size of page size a caller can request.
```

File: service/implementations/person_implementation.rb [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    finder = Person.list_in( context )
+    birth_year = context.request.list.search_data['birth_year'].to_i
+
+    unless birth_year.zero?
+      this_year_start = Date.new( birth_year )
+      next_year_start = Date.new(birth_year + 1 )
+      finder = finder.where( :date_of_birth => (this_year_start ... next_year_start))
+    end
```
Comment (davidoram) [inline service/implementations/person_implementation.rb]: Is this for [FT-816](https://loyaltynz.atlassian.net/browse/FT-816)?

File: service/models/person_model.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+class Person < Hoodoo::ActiveRecord::Base
+  # dating_enabled # Causes an issue running curl - "code": "platform.fault", "message": "PG::UndefinedTable: ERROR:  relation \"people_history_entries\" does not exist\nLINE 8:
```
Comment (davidoram) [inline service/models/person_model.rb]: Thats because when you have `dating_enabled` it requires that you install a bunch of triggers etc in the database, which we haven't looked at for this challenge. Take a look at [automatic dating](http://loyaltynz.github.io/hoodoo/guides_0300_active_record.html#heading_number_39) for more details.

**Review by davidoram — COMMENTED** [review]
Good job. Couple of very minor things in feedback. Fix them and slack me 😄

File: .byebug_history [hunk 6]
```diff
@@ -0,0 +1,239 @@
+c
```
Comment (davidoram) [inline .byebug_history]: No temporary files should be added to the git repo.  See if you can figure out how to exclude files like this one from being added to git

File: service/implementations/person_implementation.rb [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        finder = finder.where( :date_of_birth => (after_date ... long_after_date))
+      end
+    else
+      # This is the functionality for searching all records with a birth year before a given year.
+      unless before_date.zero?
+        years_ago = Date.new(birth_year - 150)
+        before_date = Date.new( birth_year )
+        finder = finder.where( :date_of_birth => (years_ago ... before_date))
+      end
+    end
```
Comment (davidoram) [inline service/implementations/person_implementation.rb]: I think this section and the logic needs a bit of work, when we look at the requirement:

> Write an rspec test that searches the database by d.o.b.  using greater than, equal to, or less than a specified date or age.

- I note that `before_date` and `after_date` are good. They meet the requirement of searching for greater than or less than a date.
- How would you search for people having an **equal to** dob?
- The requirement asks for search "greater than, equal to, or less than a specified **age**.  How would you do that in the current design? 

Here is a comment that might help you determine the logic when supplying a date rather than an age
```
# When matching peoples date of birth using dates, use the following parameters:
# 
# -  'equal_date': if supplied, then 'before_date' and 'after_date will be ignored, and only people born on that exact date will be returned.
# - 'before_date', and 'after_date'. Both are optional, either or both can be provided, so that an open ended or closed range can be supplied. For example:
# 'before_date' = 1/1/2020. means anyone born before that date
# 'after_date' = 1/1/1970. means anyone born after that date
# 'after_date' = 1/1/1970 and 'before_date' = 1/1/2020. means anyone born between those dates

File: service/implementations/person_implementation.rb [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    if before_date.zero? & after_date.zero?
+      unless birth_year.zero?
+        this_year_start = Date.new( birth_year )
+        next_year_start = Date.new(birth_year + 1 )
+        finder = finder.where( :date_of_birth => (this_year_start ... next_year_start))
+      end
+    elsif
+      # This is the functionality for searching all records with a birth year after a given year.
+      unless after_date.zero?
+        long_after_date = Date.new( birth_year + 1000 )
```
Comment (davidoram) [inline service/implementations/person_implementation.rb]: Look in the Postgres documentation to find the first and last dates it can store. Turn them into constants and use them instead

File: spec/service/implementations/person_spec.rb [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  let(:person){FactoryBot.create(:person)} # mock an instance of a person for testing
+
+  it "returns 200 if successful" do
+    delete "/1/Person/#{person.id}",
+    nil,
+    { 'CONTENT_TYPE' => 'application/json; charset=utf-8' }
+    expect(last_response.status).to eq 200
+  end
```
Comment (davidoram) [inline spec/service/implementations/person_spec.rb]: These specific tests duplicate what's in the `person_intergration_spec.rb`. 

Instead replace them with tests that target the internal workings of each endpoint.  A good thing to check in here is that the database is being updated correctly. 

For example on a `POST` check that the number of rows in the table increases by 1 etc...


```suggestion
```

File: spec/service/integration/person_intergration_spec.rb [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+    end
+
+
+
+  end
+
+  context "#update" do
+
+    context "Person exists" do
+      let(:person){FactoryBot.create(:person)} # mock an instance of a person for testing
```
Comment (davidoram) [inline spec/service/integration/person_intergration_spec.rb]: This isn't strictly speaking a 'mock' object.  

Mocks stand in for real objects, but in this case we are using FactoryBot to create 'fixtures' which is real data in the database.

Take a read about [Fixtures](https://en.wikipedia.org/wiki/Test_fixture) and  [Mocks](https://en.wikipedia.org/wiki/Mock_object)

.. and then update the comments.
### 
```suggestion
      let(:person){FactoryBot.create(:person)}  # mock an instance of a person for testing
```

File: spec/service/integration/person_intergration_spec.rb [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+          "errors": [
+            {
+              "code": "generic.invalid_parameters",
+              "message": "Body data contains unrecognised or prohibited fields",
+              "reference": "number_of_cats"
+            }
+          ],
+          "kind": "Errors"})
+        expect(last_response.body).to be_json_eql(message).excluding("interaction_id")
+      end
```
Comment (davidoram) [inline spec/service/integration/person_intergration_spec.rb]: The test doesn't match the comment in the 'it' statement.  

Maybe we don't need this test. It looks like the one earlier on line 70 `it "renders incorrectly when an addition field is present."`

File: spec/service/integration/person_intergration_spec.rb [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+      let(:person){FactoryBot.create(:person)} # mock an instance of a person for testing
+
+      it "returns 200 if successful" do
+        delete "/1/Person/#{person.id}",
+        nil,
+        { 'CONTENT_TYPE' => 'application/json; charset=utf-8' }
+        expect(last_response.status).to eq 200
+      end
+
+      it "deletes a person correctly" do
```
Comment (davidoram) [inline spec/service/integration/person_intergration_spec.rb]:
```suggestion
      it "renders the person that is deleted" do
```

File: spec/service/integration/person_intergration_spec.rb [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+
+      it "successful update http 200" do
+        # Add the mock person ID to the query
+        patch "/1/Person/#{person.id}",
+        {"name": "John"}.to_json,
+        { 'CONTENT_TYPE' => 'application/json; charset=utf-8' }
+        expect(last_response.status).to eq 200
+      end
+
+      it "updates the mock persons name" do
```
Comment (davidoram) [inline spec/service/integration/person_intergration_spec.rb]:
```suggestion
      it "renders the person, with their name updated" do
```

File: spec/service/integration/person_intergration_spec.rb [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+      let(:id) { Hoodoo::UUID.generate}# generate a random UUID to mock an id that is not in the db.
+
+      it "returns 404 if unsucessful" do
+        delete "/1/Person/#{id}",
+        nil,
+        { 'CONTENT_TYPE' => 'application/json; charset=utf-8' }
+        expect(last_response.status).to eq 404
+      end
+
+      it "deletes a person correctly" do
```
Comment (davidoram) [inline spec/service/integration/person_intergration_spec.rb]: That comment doesn't match what is being tested

File: spec/service/integration/person_intergration_spec.rb [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+          }
+        )
+        expect(last_response.body).to be_json_eql(msg).excluding("interaction_id")
+      end
+    end
+  end
+
+  context "#list" do
+
+    context "database contains multiple people" do
```
Comment (davidoram) [inline spec/service/integration/person_intergration_spec.rb]: This is great.   Can you add another contact that tests the `list` function works when the database is empty

File: spec/service/integration/person_intergration_spec.rb [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+        get "/1/Person/",
+        nil,
+        { 'CONTENT_TYPE' => 'application/json; charset=utf-8' }
+        expect(last_response.status).to eq 200
+      end
+
+      it "returns the correct message" do
+        get "/1/Person/",
+        nil,
+        { 'CONTENT_TYPE' => 'application/json; charset=utf-8' }
```
Comment (davidoram) [inline spec/service/integration/person_intergration_spec.rb]: This test is a bit *fragile* ie: It might pass sometimes, and then randomly fail - can you guess why?

File: spec/service/integration/person_intergration_spec.rb [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+      it "returns 404" do
+        get "/1/Person/#{id}",
+        nil,
+        { 'CONTENT_TYPE' => 'application/json; charset=utf-8' }
+        expect(last_response.status).to eq 404
+      end
+    end
+  end
+
+  context "#search" do
```
Comment (davidoram) [inline spec/service/integration/person_intergration_spec.rb]: I will review this section once search params code sorted

**Review by davidoram — COMMENTED** [review]
Feedback

File: spec/service/integration/person_intergration_spec.rb (hunk 11 above)
Comment (KevinWeirNZ) [reply spec/service/integration/person_intergration_spec.rb]: The test on line 70 is testing post while this one is testing patch, my thinking was that I should test that I wasn't able to update the entry with a new non-existent field. I have fixed up the comment in the it statement.

File: service/implementations/person_implementation.rb [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+    person = Person.acquire_in( context )
+    if person.nil?
+      context.response.not_found( context.request.ident )
+    else
+      context.response.set_resource( render_in( context, person ) )
+    end
+  end
+
+  # TODO in list:
+  # tidy up and refactor
```
Comment (davidoram) [inline service/implementations/person_implementation.rb]: Is this a reminder for something you still need to do?

File: service/implementations/person_implementation.rb [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+    date_of_birth_year        = context.request.list.search_data[ 'date_of_birth_year'         ].to_i
+    date_of_birth_year_after  = context.request.list.search_data[ 'date_of_birth_year_after'   ].to_i
+    date_of_birth_year_before = context.request.list.search_data[ 'date_of_birth_year_before'  ].to_i
+
+    # date of birth
+    date_of_birth             = context.request.list.search_data[ 'date_of_birth'              ].to_s
+    date_of_birth_after       = context.request.list.search_data[ 'date_of_birth_after'        ].to_s
+    date_of_birth_before      = context.request.list.search_data[ 'date_of_birth_before'       ].to_s
+
+    #minimum and maximum dates for postgres
```
Comment (davidoram) [inline service/implementations/person_implementation.rb]: Great use of Postgres min/max dates.

File: service/implementations/person_implementation.rb [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+      # Search for all entries after date of birth.
+      elsif date_of_birth.empty? & date_of_birth_before.empty? & !date_of_birth_after.empty?
+        finder = finder.where( :date_of_birth => ( Date.parse( date_of_birth_after ) ... maximum_date ) )
+      # Search for all entries between two given dates - date_of_birth_before and date_of_birth_after.
+      elsif date_of_birth.empty? & !date_of_birth_before.empty? & !date_of_birth_after.empty?
+        date_of_birth_before   = Date.parse( date_of_birth_before )
+        date_of_birth_after    = Date.parse( date_of_birth_after  )
+        finder = finder.where( :date_of_birth => ( ( date_of_birth_after - 1 ) ... ( date_of_birth_before + 1 ) ) )
+      end
+    end
```
Comment (davidoram) [inline service/implementations/person_implementation.rb]: This piece is good, but a bit hard to read.  Maybe it can be refactored / simplified a bit.  as discussed refer to https://github.com/LoyaltyNZ/hoodoo-sample/pull/6

File: spec/service/implementations/person_spec.rb [hunk 21]
```diff
@@ -0,0 +1,15 @@
+require 'spec_helper'
+
+RSpec.describe PersonImplementation do
```
Comment (davidoram) [inline spec/service/implementations/person_spec.rb:3]: Nice 👍

File: spec/service/integration/person_intergration_spec.rb [hunk 22]
```diff
@@ -0,0 +1,365 @@
+require 'spec_helper'
```
Comment (davidoram) [inline spec/service/integration/person_intergration_spec.rb]: Filename has a spelling error `person_intergration_spec.rb` -> `person_integration_spec.rb` I think :-)

File: spec/service/integration/person_intergration_spec.rb [hunk 23]
```diff
@@ -0,0 +1,365 @@
...
+        @p1 = FactoryBot.create( :person, :name => 'John 1',    :date_of_birth => '2002-03-01' )
+        @p2 = FactoryBot.create( :person, :name => 'John 2',    :date_of_birth => '2002-09-04' )
+        @p3 = FactoryBot.create( :person, :name => 'Dylan 1',   :date_of_birth => '2000-11-23' )
+        @p4 = FactoryBot.create( :person, :name => 'Dylan 2',   :date_of_birth => '1996-02-01' )
+        @p5 = FactoryBot.create( :person, :name => 'Dylan 3',   :date_of_birth => '1996-02-01' )
+      end
+
+      it "lists all entries" do
+        res = do_list('')
+        compare_lists(res, @p5, @p4, @p3, @p2, @p1)
```
Comment (davidoram) [inline spec/service/integration/person_intergration_spec.rb]: Add a `sort` order on the `PersonInterface` to make ordering explicit - see https://github.com/LoyaltyNZ/hoodoo-sample/pull/6/files

File: spec/service/integration/person_intergration_spec.rb [hunk 24]
```diff
@@ -0,0 +1,365 @@
...
+
+        it "searches for those born after 2000" do
+          res = do_list( :date_of_birth_year_after => '2000')
+          compare_lists(res, @p2, @p1)
+        end
+
+        it "searches for those born between 1996 and 2000" do
+          res = do_list( :date_of_birth_year_before => '2000', :date_of_birth_year_after => '1996' )
+
+          compare_lists(res, @p5, @p4, @p3)
```
Comment (davidoram) [inline spec/service/integration/person_intergration_spec.rb]: Seems strange that @p3 included here, but not in the test above `it "searches for those born before 2000"`

File: spec/service/integration/person_intergration_spec.rb [hunk 25]
```diff
@@ -0,0 +1,365 @@
...
+
+        it "searches for those born between 1996 and 2000" do
+          res = do_list( :date_of_birth_before => '2002-03-01', :date_of_birth_after => '1996-02-01' )
+          compare_lists(res, @p5, @p4, @p3, @p1)
+        end
+
+        # test wrong parameters
+        it "checks that date_of_birth_after is not larger than date_of_birth_before" do
+          res = do_list( :date_of_birth_after => '2002-03-01', :date_of_birth_before => '1996-02-01' )
+          expect(last_response.body).to have_json_size(0).at_path("_data")
```
Comment (davidoram) [inline spec/service/integration/person_intergration_spec.rb]: I can't remember - this might be considered an error? What did manors specs say?

File: spec/service/integration/person_intergration_spec.rb [hunk 26]
```diff
@@ -0,0 +1,365 @@
...
+        end
+
+        # test timestamps
+        it "checks that no errors occur when passing through a timestamp" do
```
Comment (davidoram) [inline spec/service/integration/person_intergration_spec.rb]: I like the way you have followed the principle of being liberal in what the API can accept - nice

**Review by davidoram — CHANGES_REQUESTED** [review]
Changed required as discussed

File: service/implementations/person_implementation.rb [hunk 27]
```diff
@@ -0,0 +1,192 @@
...
+    dob_after  = validate_date_field( context, 'date_of_birth_after'  )
+    dob_before = validate_date_field( context, 'date_of_birth_before' )
+    validate_date_range( context, dob_after, dob_before )
+
+    # Year only
+    dob_year        = validate_date_year_field( context, 'date_of_birth_year'        )
+    dob_year_after  = validate_date_year_field( context, 'date_of_birth_year_after'  )
+    dob_year_before = validate_date_year_field( context, 'date_of_birth_year_before' )
+    validate_date_range( context, dob_year_after, dob_year_before )
+    return if context.response.halt_processing?
```
Comment (davidoram) [inline service/implementations/person_implementation.rb:28]: Nice 👍

File: service/implementations/person_implementation.rb [hunk 28]
```diff
@@ -0,0 +1,192 @@
...
+          reference: { field_names: "date_of_birth" }
+        )
+      end
+    end
+    return date
+  end
+
+  def validate_date_range(context, date1, date2)
+    return if date1.nil? || date2.nil?
+      if date2 < date1
```
Comment (davidoram) [inline service/implementations/person_implementation.rb]: 🐼  This if statement block is indented too much

File: service/implementations/person_implementation.rb [hunk 29]
```diff
@@ -0,0 +1,192 @@
...
+    return finder if date.nil?
+    finder.where( 'date_of_birth::TIMESTAMP::DATE >= ?::TIMESTAMP::DATE', date)
+  end
+
+  # Year only
+  def where_dob_year_exactly( finder, date )
+    return finder if date.nil?
+    year = (date).strftime("%Y")  # Extract year from the date passed into the method
+    # Extract the year from the date in the timestamp found in the date_of_birth column and compare it to the 'year' variable.
+    finder.where( 'EXTRACT(YEAR FROM date_of_birth::TIMESTAMP::DATE) = ?', year)
```
Comment (davidoram) [inline service/implementations/person_implementation.rb:176]: Nice 💯

File: spec/service/integration/person_integration_spec.rb [hunk 30]
```diff
@@ -0,0 +1,412 @@
...
+      end
+      response = get(
+        "/1/Person#{ query }",
+        nil,
+        { 'CONTENT_TYPE' => 'application/json; charset=utf-8' })
+      expect( last_response.status).to(
+        eq( expected_code )
+      )
+      return last_response.body
+    end
```
Comment (davidoram) [inline spec/service/integration/person_integration_spec.rb]: How is this method different from `do_list`?

File: spec/service/integration/person_integration_spec.rb [hunk 31]
```diff
@@ -0,0 +1,412 @@
...
+
+      # Testing the search functionality
+      context "Search by year" do
+        it "searches for those born in 2000" do
+          res = do_list( :date_of_birth_year => '2000')
+          compare_lists(res, @p3)
+        end
+
+        it "searches for those born before 2000" do
+          res = do_list( :date_of_birth_year_before => '2000')
```
Comment (davidoram) [inline spec/service/integration/person_integration_spec.rb]: I notice that p3 not included. Is that because `date_of_birth_year_before` effectively means before the first day of that year?

**Review by davidoram — COMMENTED** [review]
A couple of small items left - good job!

File: spec/service/integration/person_integration_spec.rb (hunk 31 above)
Comment (KevinWeirNZ) [reply spec/service/integration/person_integration_spec.rb]: Yes that is right

**Review by davidoram — APPROVED** [review]
GTG 💯 👍 🥇
